### PR TITLE
fix(rewriter): rewrite accesses to document.location

### LINKED ISSUE #7402: Cross-origin error happening on same domain after click event adds same page anchor
### Current behavior:

Is not possible to continue with the test execution after getting error message:
```
Cypress detected a cross origin error happened on page load:
  > Blocked a frame with origin "http://automationpractice.com" from accessing a cross-origin frame.
Before the page load, you were bound to the origin policy:
  > http://automationpractice.com
```
This is happening when I click the create account button with a valid email (non-registered) in the email address field after visiting the site http://automationpractice.com/index.php?controller=authentication&back=my-account.

This event is not redirecting user to any new domain, actually it doesn't seem any redirection is made instead it looks like a page anchor from the same page is added to the url, loading new content in the same page and showing/hiding a few elements from the page layout.

If `chromeWebSecurity: false` is added to the the `cypress.json` file the issue continues. No error message is displayed but instead after the same event is triggered a blank page is loaded instead  blocking the test entirely.

### Desired behavior:

Cypress should be able to continue with the following step without showing the cross-origin error for this particular case.

### Test code to reproduce

1. Set up a new cypress project
2. Add a new javascript file with any desired name to the integration folder, with the following code:
```js
describe('Create Account', () => {
    it('Visit Store Website', () => {
        cy.visit('http://automationpractice.com/index.php?controller=authentication&back=my-account')
    })

    it('Add Email to create a new account', () => {
        cy.xpath('//input[@id="email_create"]').type('faketest_1@faketestexample.com')
        cy.xpath('//form[@id="create-account_form"]').submit()
    })

    it('Add personal information', () => {
        cy.get('.uniform-id_gender1').click()
        cy.get('.customer_firstname').type('First')
        cy.get('.customer_lastname').type('Last')
        cy.get('.passwd').type('whatever123')
        cy.get('.days').select('2')
        cy.get('.months').select('10')
        cy.get('.years').select('1985')
    })
})
```
3. Run the test from the cypress UI runner `npx cypress open`
4. Noticed the described issue

### Versions

* node version 12.16.3 
* npm version 6.14.4
```
"devDependencies": {
    "cypress": "^4.5.0",
    "cypress-xpath": "^1.4.0"
  }
```

### Attachments
Screen Shot 2020-05-18 at 2 07 26 PM

### PR DESCRIPTION
**Pull request by flotwig**
- Closes #7402 

### User facing changelog

- With experimentalSourceRewriting enabled, fixed unexpected behavior when using `document.location` to set the current URL.

### Additional details

* extended existing replacement for `window.location` to cover `document.location` as well
* added tests for this, a well as hash usage

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue been tagged with a release in ZenHub? 
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?
- [na] Have new configuration options been added to the [`cypress.schema.json`](cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/driver/src/cypress/resolvers.ts [hunk 1]
```diff
@@ -14,8 +14,17 @@ export function resolveWindowReference (this: typeof $Cypress, currentWindow: Wi
 
   const getTargetValue = () => {
     const targetValue = accessedObject[accessedProp]
+    const accessingDocument = dom.isDocument(accessedObject)
+    const hasLocation = dom.isWindow(accessedObject) || accessedObject
```
Comment (bahmutov) [inline packages/driver/src/cypress/resolvers.ts:18]: this is such a weird condition. is it a boolean? Probably due to `dom.isWindow`. Is it an accessed object? sure, why not. and then, if it has location and it is accessing the document, we set accessed object ... to the current window.

File: packages/driver/ts/internal-types.d.ts [hunk 2]
```diff
@@ -2,13 +2,15 @@
 // TODO: find a better place for this
 
 declare namespace Cypress {
-
   interface Cypress {
+    // TODO: how to pull these from resolvers.ts? can't import in a d.ts file...
+    resolveWindowReference: any
```
Comment (bahmutov) [inline packages/driver/ts/internal-types.d.ts:7]: nice job getting rid of all those `ts-ignore`

File: packages/driver/src/cypress/resolvers.ts (hunk 1 above)
Comment (flotwig) [reply packages/driver/src/cypress/resolvers.ts:18]: yes, it is weird :disappointed: i gave everything boolean-esque names to try and help to make the weirdness more easy to parse

File: packages/server/test/support/fixtures/projects/e2e/static/settimeout_redirect_document_href.html [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <title>Document</title>
+</head>
+<body>
+  <script type="text/javascript">
+    setTimeout(() => {
+      document.location.href = 'index.html';
+    }, 1)
```
Comment (bahmutov) [inline packages/server/test/support/fixtures/projects/e2e/static/settimeout_redirect_document_href.html:10]: why are all redirects after `setTimeout`?

Comment (flotwig) [reply packages/server/test/support/fixtures/projects/e2e/static/settimeout_redirect_document_href.html:10]: that seems to be the condition for this class of bug, it only happens when you do the setter inside of a setTimeout

File: packages/driver/ts/internal-types.d.ts (hunk 2 above)
Comment (flotwig) [reply packages/driver/ts/internal-types.d.ts:7]: yeah, as you can see i originally had good intentions of defining these by importing `resolvers.ts` and using `resolveWindowReference: typeof resolveWindowReference`, etc... but ran into an issue. know how to do the TODO?

File: packages/driver/src/cypress/resolvers.ts (hunk 1 above)
Comment (flotwig) [reply packages/driver/src/cypress/resolvers.ts]: oh i see there's a bug here too, it should be accessingDocument, not accessedObject

**Pull request merged**